File: .leetcode/15.三数之和.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 *
 * [15] 三数之和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    if( nums.length < 3 ) return [];
    let result = [];
    nums.sort((a,b)=>a-b);
    //console.log(nums);
    if( nums[0] > 0 ) return [];
    let i = 0;
    let left = 1;
    let right = nums.length - 1;
    while( i < nums.length - 2 && nums[i] <= 0){
        while(left < right){
            let sum = nums[i] + nums[left] + nums[right];
            //console.log(i,left ,right, sum);
            if( sum === 0 ){
                result.push(
                    [ nums[i], nums[left], nums[right] ]
                );
                while( left<right && nums[left+1] === nums[left] ) left++;
                while( left<right && nums[right-1] === nums[right] ) right--;
                left++;
                right--;
            }else if( sum < 0 ){
                left++;
            }else if( sum > 0 ){
                right--;
            }
        }
        while( i < nums.length - 2 && nums[i] === nums[i+1] ) i++;
        i++;
        left = i + 1;
        right = nums.length - 1;
    }
    return result;
};
// @lc code=end

//threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]);

File: leetcode/15.三数之和.go
package leetcode_15

import (
	"sort"
)

/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	if len(nums) < 3 {
		return [][]int{}
	}

	sort.Ints(nums)
	ret := [][]int{}
	var a, b, c int
	var target int
	var j, k int
	for i := 0; i < len(nums)-2; i++ {
		a = nums[i]
		if a > 0 {
			break
		}
		if i > 0 && a == nums[i-1] {
			continue
		}
		target = -a
		j = i + 1
		k = len(nums) - 1
		for j < k {
			b = nums[j]
			c = nums[k]
			if c < 0 {
				break
			}
			if b+c == target {
				ret = append(ret, []int{a, b, c})

				for j < k && nums[j+1] == b {
					j++
				}
				for j < k && nums[k-1] == c {
					k--
				}
				j++
				k--
				continue
			}
			if b+c < target {
				j++
			} else {
				k--
			}
		}
	}
	return ret
}

// @lc code=end
